### PR TITLE
Add ability to run screenshot tests on multiple devices at once

### PR DESCRIPTION
**Pull request by josedlpozo**
# I want to run screenshot tests in multiple devices

Issue: #112

### What is the goal?

The objective for this PR is to be able to run screenshot tests in multiple devices.

### How is it being implemented?

I have added a new python class called DeviceNameCalculator which is intended to calculate a unique device identifier. This is calculated using some adb commands. This device identifier will be the directory where the screenshots will be saved.

I have thought that a reliable device identifier should be based on:
- Android API Level
- Google Play Services
- Screen density
- Screen height and width

For example, a device with API level 25, Google Play Services, xxhdpi as screen density and a screen of 1920x1600 the name results as API_25_GP_XXHDPI_1920_1600. This will be the folder inside screenshots folder where the screenshots will be saved.

Screenshot filename will be the same as usual.

### How can it be tested?

For testing these new changes you can use app-example-androidjunitrunner project from examples folder.

### Required changes after this PR
Users should clear screenshot folder and run recordMode in order to create new folders based on the device which they are run.

### CODE REVIEW

**Comment by josedlpozo** [comment]
@tdrhq @xiphirx I have closed the another PR and opened this one with the changes required. I wait for your news 😃 

Thank u!

**Comment by j3susalonso** [comment]
Any updates?? Seems like a nice feature to have... @tdrhq @xiphirx

**Comment by xiphirx** [comment]
I will get to this PR soon, it changes a lot of things and I'm currently busy syncing things in our internal repo.

File: examples/app-example-androidjunitrunner/app/src/androidTest/java/com/example/screenshots/MainActivityTest.java [hunk 1]
```diff
@@ -21,21 +21,24 @@
 public class MainActivityTest {
 
     @Rule
-    public ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(MainActivity.class);
+    public ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(MainActivity.class, true, false);
```
Comment (xiphirx) [inline examples/app-example-androidjunitrunner/app/src/androidTest/java/com/example/screenshots/MainActivityTest.java]: Why the change here?

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+class DeviceNameCalculator:
+    def __init__(self):
+        pass
+
+    def name(self):
+        api_version_text = self._api_version_text()
+        play_services_text = self._play_services_text()
+        screen_density_text = self._screen_density_text()
+        screen_size_text = self._screen_size_text()
+
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: We should ensure that we have valid values here, and not `None`. If we receive `None` from any of the above functions then something is broken, so we should throw an exception here.

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return "{0}_{1}_{2}_{3}".format(api_version_text, play_services_text,
+                                        screen_density_text, screen_size_text)
+
+    def _screen_density_text(self):
+        density = self._screen_density()
+
+        try:
+            density = int(density)
+        except:
+            return 'UNKNOWN'
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: Instead of defaulting to UNKNOWN I would rather let this exception cause an error. `_screen_density` only looks for a contiguous string of numbers, so this _should_ never fail. In case it does, we shouldn't continue, especially if `None` was returned.

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            return 'LDPI'
+        elif density in range(121, 161):
+            return 'MDPI'
+        elif density in range(161, 241):
+            return 'HDPI'
+        elif density in range(241, 321):
+            return 'XHDPI'
+        elif density in range(321, 481):
+            return 'XXHDPI'
+        else:
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: nit: this else is unnecessary

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        if density:
+            return density.group(0)
+
+    def _screen_size(self):
+        result = common.check_output([common.get_adb()] + ['shell', 'wm', 'size'])
+        density = re.search('[0-9]+x[0-9]+', result)
+        if density:
+            return density.group(0)
+
+    def _screen_size_text(self):
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: Not sure if this needs to be a function.

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    def _api_version(self):
+        return common.check_output(
+            [common.get_adb()] + ['shell', 'getprop', 'ro.build.version.sdk'])
+
+    def _api_version_text(self):
+        api_version = ''.join(self._api_version().splitlines())
+        return 'API_' + api_version
+
+    def _play_services_text(self):
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: Lets move this function up near `_has_play_services`

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        density = re.search('[0-9]+x[0-9]+', result)
+        if density:
+            return density.group(0)
+
+    def _screen_size_text(self):
+        return self._screen_size()
+
+    def _has_play_services(self):
+        output = common.check_output(
+            [common.get_adb()] + ['shell', 'pm', 'dump', 'com.google.android.gms', '|', 'grep',
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: We should check `adb shell pm path com.google.android.gms` instead. Less output and the same information can be gathered from the output. You will also need to catch an exception here when adb returns a non zero exit code.

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py (hunk 2 above)
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: We should include the architecture in the name.

File: plugin/src/py/android_screenshot_tests/device_name_calculator.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                                  'version'])
+        return True if output else False
+
+    def _api_version(self):
+        return common.check_output(
+            [common.get_adb()] + ['shell', 'getprop', 'ro.build.version.sdk'])
+
+    def _api_version_text(self):
+        api_version = ''.join(self._api_version().splitlines())
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/device_name_calculator.py]: `api_version = int(self._api_version())` ?

File: examples/app-example-androidjunitrunner/app/src/androidTest/java/com/example/screenshots/DetailActivityTest.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+  private Intent givenAnOkText() {
+    Intent intent = new Intent();
+    intent.putExtra(DetailActivity.TYPE,
+        DetailActivity.Type.OK);
+    intent.putExtra(DetailActivity.TEXT,
+        "Screenshot testing is wonderful");
+
+    return intent;
+  }
```
[the remark this answers is not in the document]
Comment (josedlpozo) [reply examples/app-example-androidjunitrunner/app/src/androidTest/java/com/example/screenshots/DetailActivityTest.java]: Done 👍

File: examples/app-example-androidjunitrunner/app/src/androidTest/java/com/example/screenshots/MainActivityTest.java (hunk 1 above)
Comment (josedlpozo) [reply examples/app-example-androidjunitrunner/app/src/androidTest/java/com/example/screenshots/MainActivityTest.java]: We need this change if we want to launch the activity the way we do it now. If we do as before, mainActivityTestSettingsOpen is not a deterministic test.

**Comment by josedlpozo** [comment]
Hi @xiphirx! I have already updated the code with your suggestions, please let me know if you have any more doubts!

**Comment by xiphirx** [comment]
It seems like you need to re-record the screenshots since the device name calculator takes architecture into account now.

**Comment by josedlpozo** [comment]
Yeap, I forgot it. I have already included them!

File: plugin/src/py/android_screenshot_tests/pull_screenshots.py [hunk 10]
```diff
@@ -445,7 +450,8 @@ def main(argv):
                             opt_generate_png=opts.get('--generate-png'),
                             record=opts.get('--record'),
                             verify=opts.get('--verify'),
-                            adb_puller=SimplePuller(puller_args))
+                            adb_puller=SimplePuller(puller_args),
+                            device_name_calculator=DeviceNameCalculator())
```
Comment (tdrhq) [inline plugin/src/py/android_screenshot_tests/pull_screenshots.py]: you mentioned that you're not using a DeviceNameCalculator by default, but this seems to suggest otherwise. Am I missing something?

Comment (josedlpozo) [reply plugin/src/py/android_screenshot_tests/pull_screenshots.py]: where I have mentioned that I'm not using a DeviceNameCalculator by default? I look at this PR and I can't see it.
I think that you are talking about this PR https://github.com/facebook/screenshot-tests-for-android/pull/113, which is closed because you requested some changes and this one was created with the changes instead of changing that.

Did you have a look at this one?
Do you have any news?

Comment (tdrhq) [reply plugin/src/py/android_screenshot_tests/pull_screenshots.py]: okay yeah, we definitely need this to default to none, otherwise it's going to break existing users directory structures (and definitely some tools internally at FB :) ). The new format can be made opt-in, perhaps in the future we can turn it on by default if enough people think it's the right move 

(@xiphirx is doing most of the review, I was just scanning through this for obvious breakages)

Comment (josedlpozo) [reply plugin/src/py/android_screenshot_tests/pull_screenshots.py]: I was thinking about a NoOpDeviceNameCalculator which returns an empty string when name method is called. This way we don't work with None's. What do you think?

By default, NoOpDeviceNameCalculator is used but users can turn DeviceNameCalculator by setting a true multipleDevices variable in their build.gradle file.

What do you think about creating a new app example project with this multiple devices feature turned on?

Is it okey for you, @xiphirx @tdrhq?

Comment (tdrhq) [reply plugin/src/py/android_screenshot_tests/pull_screenshots.py]: Yep, NoOpDeviceNameCalculator sounds like the right design (but: it might need some more changes because the device_name_calculator right now assumes that it's going into a subdirectory, but the default mode shouldn't go into a subdirectory)

new example app: that would be fantastic, go for it

Comment (josedlpozo) [reply plugin/src/py/android_screenshot_tests/pull_screenshots.py]: Hey! I've updated the PR with the changes requested. I took the freedom to write the example app in Kotlin.

About this 'device_name_calculator right now assumes that it's going into a subdirectory, but the default mode shouldn't go into a subdirectory', I think that you're talking about the change in recorder.py from os.mkdir to os.makedirs. Makedirs works also if you want to create a just one directory nor only with multiple directories.

@xiphirx @tdrhq If you have any news, please let me know!

**Comment by xiphirx** [comment]
Hey @josedlpozo it seems like there are some merge conflicts that need to be resolved.

**Comment by josedlpozo** [comment]
Hey @xiphirx conflicts are resolved. Do you have any news??

**Comment by xiphirx** [comment]
Hi @josedlpozo I'm currently trying to import your code internally, it takes a bit to do so. After, I need to test a few things and then I should be able to merge it in.

File: examples/app-example-multipledevices/app/proguard-rules.pro [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+#   public *;
+#}
+
+# Uncomment this to preserve the line number information for
+# debugging stack traces.
+#-keepattributes SourceFile,LineNumberTable
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
```
Comment (xiphirx) [inline examples/app-example-multipledevices/app/proguard-rules.pro]: Please remove this file.

File: examples/app-example-multipledevices/app/src/main/java/com/facebook/screenshots/MessageType.kt [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.facebook.screenshots
+
+import java.io.Serializable
+
+sealed class MessageType : Serializable {
```
Comment (xiphirx) [inline examples/app-example-multipledevices/app/src/main/java/com/facebook/screenshots/MessageType.kt]: Lets not implement Serializable here and instead just use plain `Int`s

File: examples/app-example-multipledevices/app/src/test/java/com/facebook/screenshots/ExampleUnitTest.kt [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
+    @Test
+    fun addition_isCorrect() {
+        assertEquals(4, 2 + 2)
+    }
+}
```
Comment (xiphirx) [inline examples/app-example-multipledevices/app/src/test/java/com/facebook/screenshots/ExampleUnitTest.kt]: Please delete this file.

File: plugin/src/py/android_screenshot_tests/pull_screenshots.py [hunk 14]
```diff
@@ -43,10 +47,12 @@
 KEY_WIDTH = 'width'
 KEY_HEIGHT = 'height'
 
+
 def usage():
-    print("usage: ./scripts/screenshot_tests/pull_screenshots com.facebook.apk.name.tests [--generate-png]", file=sys.stderr)
+    print ( "usage: ./scripts/screenshot_tests/pull_screenshots com.facebook.apk.name.tests [--generate-png]", file=sys.stderr)
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/pull_screenshots.py]: revert this please

File: plugin/src/py/android_screenshot_tests/pull_screenshots.py [hunk 15]
```diff
@@ -171,7 +180,7 @@ def write_view_hierarchy_overlay_nodes(hierarchy, html, parent_id):
     while not to_output.empty():
         node = to_output.get()
         left = node[KEY_LEFT]
-        top = node[KEY_TOP]
+        top  = node[KEY_TOP]
```
Comment (xiphirx) [inline plugin/src/py/android_screenshot_tests/pull_screenshots.py]: revert this please
